File: solutions/Leetcode_877/leetcode_877.py
from typing import List


class Solution1:
    def stone_game(self, piles: List[int]) -> bool:
        num_piles = len(piles)
        dp = [[0] * num_piles for _ in range(num_piles)]
        for d in range(1, num_piles):
            for i in range(num_piles-d):
                dp[i][i+d] = max(piles[i] - dp[i+1][i+d], piles[i+d] - dp[i][i+d-1])
        return dp[0][-1] > 0

File: solutions/Leetcode_877/leetcode_877.cpp
#include <vector>


class Solution1
{
public:
    bool stoneGame(std::vector<int>& piles)
    {
        int num_piles = piles.size();
        std::vector<std::vector<int>> dp(num_piles, std::vector<int>(num_piles, 0));
        for (int d = 1; d < num_piles; d++)
            for (int i = 0; i < num_piles - d; i++)
                dp[i][i+d] = std::max(piles[i] - dp[i+1][i+d], piles[i+d] - dp[i][i+d-1]);
        return dp[0][num_piles-1];
    }
};
